node/core/backing/src/lib.rs

```diff
@@ -298,21 +323,272 @@ fn table_attested_to_backed(
+
+	tx_from.send(FromJob::CandidateValidation(
+			CandidateValidationMessage::ValidateFromChainState(
+				candidate,
+				pov,
+				tx,
+			)
+		)
+	).await?;
+
+	Ok(rx.await??)
+}
+
+type BackgroundValidationResult = Result<(CandidateReceipt, CandidateCommitments, Arc<PoV>), CandidateReceipt>;
+
+struct BackgroundValidationParams<F> {
+	tx_from: mpsc::Sender<FromJob>,
+	tx_command: mpsc::Sender<ValidatedCandidateCommand>,
+	candidate: CandidateReceipt,
+	relay_parent: Hash,
+	pov: Option<Arc<PoV>>,
+	validator_index: Option<ValidatorIndex>,
+	n_validators: usize,
+	make_command: F,
+}
+
+async fn validate_and_make_available(
+	params: BackgroundValidationParams<impl Fn(BackgroundValidationResult) -> ValidatedCandidateCommand>,
+) -> Result<(), Error> {
+	let BackgroundValidationParams {
+		mut tx_from,
+		mut tx_command,
+		candidate,
+		relay_parent,
+		pov,
+		validator_index,
+		n_validators,
+		make_command,
+	} = params;
+
+	let pov = match pov {
+		Some(pov) => pov,
+		None => request_pov_from_distribution(
+			&mut tx_from,
+			relay_parent,
+			candidate.descriptor.clone(),
+		).await?,
+	};
+
+	let v = request_candidate_validation(&mut tx_from, candidate.descriptor.clone(), pov.clone()).await?;
+
+	let expected_commitments_hash = candidate.commitments_hash;
+
+	let res = match v {
+		ValidationResult::Valid(commitments, validation_data) => {
+			// If validation produces a new set of commitments, we vote the candidate as invalid.
+			if commitments.hash() != expected_commitments_hash {
+				Err(candidate)
+			} else {
+				let erasure_valid = make_pov_available(
+					&mut tx_from,
+					validator_index,
+					n_validators,
+					pov.clone(),
+					candidate.hash(),
+					validation_data,
+					candidate.descriptor.erasure_root,
+				).await?;
+
+				match erasure_valid {
+					Ok(()) => Ok((candidate, commitments, pov.clone())),
+					Err(InvalidErasureRoot) => Err(candidate),
+				}
+			}
+		}
+		ValidationResult::Invalid(_reason) => {
+			Err(candidate)
+		}
+	};
+
+	let command = make_command(res);
+	tx_command.send(command).await?;
+	Ok(())
+}
+
 impl CandidateBackingJob {
 	/// Run asynchronously.
 	async fn run_loop(mut self) -> Result<(), Error> {
-		while let Some(msg) = self.rx_to.next().await {
-			match msg {
-				ToJob::CandidateBacking(msg) => {
-					self.process_msg(msg).await?;
+		loop {
+			futures::select! {
+				validated_command = self.background_validation.next() => {
+					if let Some(c) = validated_command {
+						self.handle_validated_candidate_command(c).await?;
+					} else {
+						panic!("`self` hasn't dropped and `self` holds a reference to this sender; qed");
+					}
+				}
+				to_job = self.rx_to.next() => match to_job {
+					None | Some(ToJob::Stop) => break,
+					Some(ToJob::CandidateBacking(msg)) => {
+						self.process_msg(msg).await?;
+					}
+				}
+			}
+		}
+
+		Ok(())
+	}
+
+	#[tracing::instrument(level = "trace", skip(self, command), fields(subsystem = LOG_TARGET))]
```

Comment:
If we `impl Debug for ValidateCandidateCommand`, then we don't have to skip `command` here. Obviously it's not that useful to include the entire `BackgroundValidationResult` in the debug output, but emitting at least the name of the variant could improve the utility of the logging output here.